src/clp_ffi_js/ir/UnstructuredIrStreamReader.hpp

```diff
@@ -0,0 +1,85 @@
+#ifndef CLP_FFI_JS_IR_UNSTRUCTUREDIRSTREAMREADER_HPP
+#define CLP_FFI_JS_IR_UNSTRUCTUREDIRSTREAMREADER_HPP
+
+#include <Array.hpp>
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include <clp/ir/types.hpp>
+#include <clp/TimestampPattern.hpp>
+#include <emscripten/val.h>
+
+#include <clp_ffi_js/ir/LogEventWithLevel.hpp>
+#include <clp_ffi_js/ir/StreamReader.hpp>
+#include <clp_ffi_js/ir/StreamReaderDataContext.hpp>
+
+namespace clp_ffi_js::ir {
+using clp::ir::four_byte_encoded_variable_t;
+
+/**
+ * Mapping between an index in the filtered log events collection to an index in the unfiltered
+ * log events collection.
+ */
+using FilteredLogEventsMap = std::optional<std::vector<size_t>>;
+
+/**
+ * Class to deserialize and decode Zstd-compressed CLP unstructured IR streams, as well as format
+ * decoded log events.
+ */
+class UnstructuredIrStreamReader : public StreamReader {
+public:
+    // Destructor
+    ~UnstructuredIrStreamReader() override = default;
+
+    // Disable copy constructor and assignment operator
+    UnstructuredIrStreamReader(UnstructuredIrStreamReader const&) = delete;
+    auto operator=(UnstructuredIrStreamReader const&) -> UnstructuredIrStreamReader& = delete;
+
+    // Define default move constructor
+    UnstructuredIrStreamReader(UnstructuredIrStreamReader&&) = default;
+    // Delete move assignment operator since it's also disabled in `clp::ir::LogEventDeserializer`.
+    auto operator=(UnstructuredIrStreamReader&&) -> UnstructuredIrStreamReader& = delete;
+
+    /**
+     * First packages a `StreamReaderDataContext` using inputs and a unstructured IR deserializer,
+     * then creates a `UnstructuredIrStreamReader`.
+     *
+     * @param zstd_decompressor
+     * @param data_array An array containing a Zstandard-compressed IR stream.
```

Comment:
```suggestion
     * @param zstd_decompressor A decompressor for an IR stream, where the read head of the stream
     * is just after the stream's encoding type.
     * @param data_array The array backing `zstd_decompressor`.
```

I deleted the description because I think it's mostly implementation details. I think the important parts for the caller are what I added in the parameter descriptions.